worlds/pokemon_crystal/rom.py

```diff
@@ -0,0 +1,425 @@
+
+    if world.options.better_marts:
+        mart_address = data.rom_addresses["Marts"]
+        better_mart_address = data.rom_addresses["MartBetterMart"] - 0x10000
+        better_mart_bytes = better_mart_address.to_bytes(2, "little")
+        for i in range(33):
+            # skip goldenrod and celadon
+            if i not in [6, 7, 8, 9, 10, 11, 12, 24, 25, 26, 27, 28]:
+                write_bytes(patch, better_mart_bytes, mart_address)
+            mart_address += 2
+
+    hmbadges_address = data.rom_addresses["AP_Setting_HMBadges"] + 1
+    write_bytes(patch, [world.options.hm_badge_requirements.value], hmbadges_address)
+
+    exp_modifier_address = data.rom_addresses["AP_Setting_ExpModifier"] + 1
+    write_bytes(patch, [world.options.experience_modifier], exp_modifier_address)
+
+    elite_four_text = convert_to_ingame_text("{:02d}".format(world.options.elite_four_badges.value))
+    write_bytes(patch, elite_four_text, data.rom_addresses["AP_Setting_VictoryRoadBadges_Text"] + 1)
+    write_bytes(patch, [world.options.elite_four_badges - 1],
+                data.rom_addresses["AP_Setting_VictoryRoadBadges"] + 1)
+    rocket_badges = world.options.elite_four_badges - 2 if world.options.elite_four_badges > 1 else 0
+    write_bytes(patch, [rocket_badges], data.rom_addresses["AP_Setting_RocketBadges"] + 1)
+
+    red_text = convert_to_ingame_text("{:02d}".format(world.options.red_badges.value))
+    write_bytes(patch, red_text, data.rom_addresses["AP_Setting_RedBadges_Text"] + 1)
+    write_bytes(patch, red_text, data.rom_addresses["AP_Setting_RedBadges_Text2"] + 1)
+    write_bytes(patch, [world.options.red_badges - 1], data.rom_addresses["AP_Setting_RedBadges_Oak"] + 1)
+    write_bytes(patch, [world.options.red_badges - 1], data.rom_addresses["AP_Setting_RedBadges_Gate"] + 1)
+
+    trainersanity_alerts_address = data.rom_addresses["AP_Setting_TrainersanityMessages"] + 1
+    write_bytes(patch, [world.options.trainersanity_alerts], trainersanity_alerts_address)
+
+    for i, script in enumerate(world.generated_phone_traps):
+        address = data.rom_addresses["AP_Setting_PhoneCallTrapTexts"] + (i * 0x400)
+        s_bytes = script.get_script_bytes()
+        # write script text
+        write_bytes(patch, s_bytes, address)
+        # write script caller id
+        address = data.rom_addresses["AP_Setting_SpecialCalls"] + (6 * i) + 2
+        write_bytes(patch, [script.caller_id], address)
+
+    phone_location_bytes = []
+    for loc in world.generated_phone_indices:
+        phone_location_bytes += list(loc.to_bytes(2, "little"))
+    phone_location_address = data.rom_addresses["AP_Setting_Phone_Trap_Locations"]
+    write_bytes(patch, phone_location_bytes, phone_location_address)
+
+    start_inventory_address = data.rom_addresses["AP_Start_Inventory"]
+    start_inventory = copy.deepcopy(world.options.start_inventory.value)
+    for item, quantity in start_inventory.items():
```

Comment:
I'd recommend building this from `world.multiworld.precollected_items[world.player]` instead. If you ever decide to automatically add something into your players' starting inventories, it's usually recommended to use `push_precollected` instead of modifying the player's options. And you want to give the player anything that the multiworld as a whole has decided is precollected, which may be partially outside your control. The generator's relatively new `panic_method` setting might put unplaced progression items into `precollected_items`. I've not tried that setting out myself though.